cmd/snap-confine/snap-confine.c

```diff
@@ -894,7 +938,83 @@ static void enter_non_classic_execution_environment(sc_invocation *inv, struct s
+        getresgid(&rgid, &egid, &sgid);
+
+        /* We are going to change uid or gid so let's give us the capability to
+         * do so and set "keep caps" so that we are not going to lose
+         * capabilities after switching back to non-root user. */
+        if ((ruid != 0 || euid != 0 || suid != 0) || (rgid != 0 || egid != 0 || sgid != 0)) {
+            /* Once we switch back to the non-root user, we will drop the
+             * effective capabilities but this is okay because after this
+             * function returns we will again manipulate capabilities before
+             * the next privileged operation. */
+            if (prctl(PR_SET_KEEPCAPS, 1) < 0) {
+                die("cannot set PR_SET_KEEPCAPS 1");
+            }
+            cap_t c SC_CLEANUP(sc_cleanup_cap_t) = cap_get_proc();
+            if (c == NULL) {
+                die("cannot obtain current caps");
+            }
+            if (cap_set_flag(c, CAP_EFFECTIVE, SC_ARRAY_SIZE(snap_confine_caps_extra_cgroup_v1),
+                             snap_confine_caps_extra_cgroup_v1, CAP_SET) != 0) {
+                die("cannot fill effective capability set (setuid)");
+            }
+            if (cap_set_proc(c) != 0) {
+                die("cannot set privileged capabilities");
+            }
+            sc_debug_capabilities("after cap_set_proc");
+        }
+
+        if (ruid != 0 || euid != 0 || suid != 0) {
+            if (setresuid(0, 0, 0) < 0) {
+                die("cannot setresuid(%d, %d, %d) before freezer join", 0, 0, 0);
+            }
+        }
+        if (rgid != 0 || egid != 0 || sgid != 0) {
+            if (setresgid(0, 0, 0) < 0) {
+                die("cannot setresgid(%d, %d, %d) before freezer join", 0, 0, 0);
+            }
+        }
+        sc_debug_capabilities("caps at freezer join");
         sc_cgroup_freezer_join(inv->snap_instance, getpid());
+        if (rgid != 0 || egid != 0 || sgid != 0) {
+            if (setresgid(rgid, egid, sgid) < 0) {
+                die("cannot setresgid(%d, %d, %d) after freezer join", rgid, egid, sgid);
+            }
+        }
+        if (ruid != 0 || euid != 0 || suid != 0) {
+            if (setresuid(ruid, euid, suid) < 0) {
+                die("cannot setresuid(%d, %d, %d) after freezer join", ruid, euid, suid);
+            }
+        }
+        /* We no longer need CAP_SETUID or CAP_SETGID in the permitted set.
+         * Let's drop them as a hardening measure. */
+        cap_t c SC_CLEANUP(sc_cleanup_cap_t) = cap_get_proc();
+        if (c == NULL) {
+            die("cannot obtain current caps");
+        }
+        if (cap_set_flag(c, CAP_PERMITTED, SC_ARRAY_SIZE(snap_confine_caps_extra_cgroup_v1),
+                         snap_confine_caps_extra_cgroup_v1, CAP_CLEAR) != 0) {
+            die("cannot clear permitted capability set (setuid,setgid)");
+        }
+        if (cap_set_proc(c) != 0) {
+            die("cannot set reduced capabilities (setuid,setgid)");
+        }
```

Comment:
can't we do this unconditionally also for v2? we are doing after the function, but could do this bit earlier?

Reply:
I missed https://github.com/canonical/snapd/pull/15743/files#diff-01856638ab88ad8d386853fa8a00e8422d1688ccbb852dcc43973f650352017aR464 on my last read